qualang_tools/plot/Fitting/fitting.py

```diff
@@ -0,0 +1,418 @@
+from scipy import optimize
+from sklearn import preprocessing
+import matplotlib.pyplot as plt
+import itertools
+import json
+import numpy as np
+
+
+class Fit:
+    """
+    This class takes care of the fitting to the measured data.
+    All the fitting functions are taken from the `Fitting_functions` file.
+    It includes:
+        - Fitting to a linear line
+        - Fitting to Ramsey experiment
+        - Fitting to transmission resonator spectroscopy
+        - Fitting to reflection resonator spectroscopy
+
+    """
+
+    @staticmethod
+    def linear(x, y):
+        """
+           Create a linear fit of the form
+
+           .. math::
+           f(x) = a * x + b
+
+           for unknown parameters :
+                a - The slope of the function
+                b - The free parameter of the function
+
+
+           :param x: The data on the x axis
+           :param y: The data on the y axis
+           :return: A dictionary of (fit_func, a, b)
+
+           """
+
+        # Normalizing the vectors
+        xn = preprocessing.normalize([x], return_norm=True)
+        yn = preprocessing.normalize([y], return_norm=True)
+        x = xn[0][0]
+        y = yn[0][0]
+
+        # Finding an initial guess to the slope
+        m0 = (y[-1] - y[0]) / (x[-1] - x[0])
+
+        # Finding an initial guess to the free parameter
+        n0 = y[0]
+
+        fit_type = lambda x, a: m0 * a[0] * x + n0 * a[1]
+
+        def curve_fit(f, x, y, a0):
+            def opt(x, y, a):
+                return np.sum(np.abs(f(x, a) - y) ** 2)
+
+            out = optimize.minimize(lambda a: opt(x, y, a), a0)
+            if not out.success: raise Exception(
+                "The fitting failed, please check the quality of the data or the fitting function that you are using")
+            return out["x"]
+
+        popt = curve_fit(fit_type, x, y, [m0, n0])
+
+        print(
+            f"a = {popt[0] * yn[1] / xn[1]}, b = {popt[1] * yn[1]}"
+        )
+        out = {
+            "fit_func": lambda x: fit_type(x, popt) * yn[1],
+            "a": popt[0] * yn[1] / xn[1],
+            "b": popt[1] * yn[1],
+        }
+
+        return out
+
+    @staticmethod
+    def ramsey(x, y):
```

Comment:
I modified the program quite a lot to make it more robust, readable and with uncertainty in the fitted parameters:
`def ramsey(x, y, verbose=0):
    """
      Create a fit to Ramsey experiment of the form

      .. math::
      f(x) = final_offset * (1 - np.exp(-x * (1/T2))) + amp / 2 * (
          np.exp(-x * (1/tau))
          * (initial_offset * 2 + np.cos(2 * np.pi * f * x + phase))
          )

      for unknown parameters :
          f - The detuning frequency [GHz]
          phase - The phase [rad]
          T2 - The decay constant [ns]
          amp - The amplitude [a.u.]
          final_offset -  The offset visible for long dephasing times [a.u.]
          initial_offset - The offset visible for short dephasing times

      :param x: The dephasing time [ns]
      :param y: Data containing the Ramsey signal
      :param verbose: if 1 prints data into the terminal
      :return: A dictionary of (fit_func, f, phase, tau, amp, uncertainty_population, initial_offset)

       """

    # Normalizing the vectors
    xn = preprocessing.normalize([x], return_norm=True)
    yn = preprocessing.normalize([y], return_norm=True)
    x = xn[0][0]
    y = yn[0][0]

    # Compute the FFT for guessing the frequency
    fft = np.fft.fft(y)
    f = np.fft.fftfreq(len(x))
    # Take the positive part only
    fft = fft[1: len(f)//2]
    f = f[1: len(f)//2]
    # Remove the DC peak if there is one
    if (np.abs(fft)[1:] - np.abs(fft)[:-1] > 0).any():
        first_read_data_ind = np.where(np.abs(fft)[1:] - np.abs(fft)[:-1] > 0)[0][0]  # away from the DC peak
        fft = fft[first_read_data_ind:]
        f = f[first_read_data_ind:]

    # Finding a guess for the frequency
    out_freq = f[np.argmax(np.abs(fft))]
    guess_freq = out_freq / (x[1] - x[0])

    # The period is 1 / guess_freq --> number of oscillations --> peaks decay to get guess_T2
    period = int(np.ceil(1 / out_freq))
    peaks = np.array([np.std(y[i * period: (i + 1) * period]) for i in range(round(len(y) / period))]) * np.sqrt(2) * 2

    # Finding a guess for the decay (slope of log(peaks))
    if len(peaks) > 1:
        guess_T2 = -1 / ((np.log(peaks)[-1] - np.log(peaks)[0]) / (period * (len(peaks) - 1))) * (x[1] - x[0])
    else:
        guess_T2 = 100 / xn[1][0]
        print(Warning("WARNING: The initial guess for the decay failed, increasing the number of oscillations should solve the issue."))

    # Finding a guess for the offsets
    initial_offset = np.mean(y[:period])
    post_decay_mean = np.mean(y[-period:])

    # Finding a guess for the phase
    guess_phase = np.angle(fft[np.argmax(np.abs(fft))]) - guess_freq * 2 * np.pi * x[0]

    # Fitting function
    fit_type = lambda x, a: post_decay_mean * a[4] * (1 - np.exp(-x / guess_T2 * a[1])) + peaks[0] / 2 * a[2] * (
            np.exp(-x / guess_T2 * a[1])
            * (a[5] * initial_offset / peaks[0] * 2 + np.cos(2 * np.pi * a[0] * guess_freq * x + a[3]))
    )
    if verbose:
        print(
            f"Initial guess:\n f = {guess_freq / xn[1][0]}, phase = {guess_phase}, T2 = {guess_T2 * xn[1][0]}, amp = {peaks[0] * yn[1][0]}, uncertainty population = {post_decay_mean * yn[1][0]}, initial offset = {initial_offset * yn[1][0]}"
        )
    import scipy
    def func(x, a0, a1, a2, a3, a4, a5):
        return post_decay_mean * a4 * (1 - np.exp(-x / guess_T2 * a1)) + peaks[0] / 2 * a2 * (
                np.exp(-x / guess_T2 * a1)
                * (a5 * initial_offset / peaks[0] * 2 + np.cos(2 * np.pi * a0 * guess_freq * x + a3))
        )

    popt, pcov = scipy.optimize.curve_fit(
        func,
        x,
        y,
        p0=[1, 1, 1, guess_phase, 1, 1],
    )

    perr = np.sqrt(np.diag(pcov))

    out = {
        "fit_func": lambda x: fit_type(x/xn[1][0], popt) * yn[1][0],
        "f": popt[0] * guess_freq / xn[1][0],
        "phase": popt[3] % (2 * np.pi),
        "T2": (guess_T2 * popt[1]) * xn[1][0],
        "amp": peaks[0] * popt[2] * yn[1][0],
        "uncertainty_population": post_decay_mean * popt[4] * yn[1][0],
        "initial_offset": popt[5] * initial_offset * yn[1][0],
    }
    if verbose:
        print(
            f"Fit results:\n"
            f"f = {out['f']*1000:.3f} +/- {perr[0] * guess_freq / xn[1][0]:.3f} MHz, \n"
            f"phase = {out['phase']:.3f} +/- {perr[3] % (2 * np.pi):.3f} rad, \n"
            f"T2 = {out['T2']:.2f} +/- {perr[1] * guess_T2 * xn[1][0]:.3f} ns, \n"
            f"amp = {out['amp']:.2f} +/- {perr[2] * peaks[0] * yn[1][0]:.3f} a.u., \n"
            f"uncertainty population = {out['uncertainty_population']:.2f} +/- {perr[4] * post_decay_mean * yn[1][0]:.3f} a.u., \n"
            f"initial offset = {out['initial_offset']:.2f} +/- {perr[5] * initial_offset * yn[1][0]:.3f}"
        )
    return out`

We can discuss it whenever you want